### PR TITLE
Major Update to Work With 2.3.1 (latest)

### PR DESCRIPTION
**Pull request by nmaludy**
## Intro
Lots of work has been done here and i'll try to summarize as best i can. 

## Overview
When starting work on this module it failed to install numerous components in the `profiles` directory. Most of this was due to failures in external packages (rabbitmq, mongodb, postgres, etc), dependency/ordering issues in the code (you'll see lots of updates to with -> ), and/or changes to the st2 code since the module was last updated. 

### manifests/auth/standalone.pp

- D not have access to the `::st2` variables.
- Had a dependency issue where (on some platforms) allowed the `htpasswd` file to be created after the st2 services were starting
- Created an unnecessary "test user"

### manifests/auth_user.pp

- Dependency issues here where the `htpasswd` file was sometimes trying to be created before the `/etc/st2` directory was created, and other times it was trying to be created after the st2 services had started.

### manifests/init.pp

- Needed extra variables for SSL setup in st2web. 
- Needed extra variables for proper database setup (mongodb and postgres)
- Needed path to the st2auth logging config file
- Needed variables about the datastore encryption keys

### manifests/kv.pp

- Some puppet lint problems (notice the whitespace fix and reordering of class params)
- Dependency issues where the tag being used for the `Service` resource was incorrect
- Dependency issues where sometimes st2 hadn't been reloaded so the k/v loads would fail

### manifests/notices.pp

- Puppet lint fixes for using double quotes without variable interpolation in the string.

### manifests/pack.pp

- Unit tests revealed that many of the dependencies of this resource were not declared (group and directories)
- Pointing at old location for config directory
- Needed lots of dependency work to ensure resources were created in the proper order

### manifests/params.pp

- Broke down the old `st2_server_packages` variable into various components to align more with what ansible-st2 and the "one liner" shell scripts do in their functions.
- Removed some unused code in the "init provider" section
- Broke down the old `st2_services` into its components similar to `st2_server_packages`. FYI: The mistral services are handled by the mistral install instead of being grouped together into `st2 server`.
- Added lots of new parameters for services that were not configured in the past like (nginx, st2web, mongodb, rabbitmq)

### manifests/profile/client.pp

- Removed stale comment

### manifests/profile/fullinstall.pp

- Mainly dependency cleanup here.
- Ensure that packages are installed in the correct order and that there are meaningful anchors in place in case others need to execute tasks at certain points during the install.

### manifests/profile/mistral.pp

- This was completely re-written
- Previously it was performing a lot of tasks manually that i believe st2mistral package now handles for us

### manifests/profile/mongodb.pp

- Completely re-written
- It now handles auth (did not previously)
- It also deals with several deficiencies in the puppetlabs-mongodb module. This module has lots of annoying bugs. I'm not at the point where i want to code up a new module myself yet, but we do have to work around several quirks for this to even work (sorry!).

### manifests/profile/nginx.pp

- New profile that installs and configures nginx (does not setup st2web config, that is left to the st2web profile)
- Utilizes the nginx puppet module to do all of the heavy lifting here

### manifests/profile/nodejs.pp

- Completely re-written
- Utilizes the nodejs puppet module to do all of the heavy lifting instead of doing it ourselves
- Works around a small quirk of the module on RedHat distributions

### manifests/profile/postgresql.pp

- Expanded this to properly configure postgres for listening according to the standard installs (shell scripts and ansible-st2)
- Also ensured that 9.4 is installed on RHEL6

### manifests/profile/rabbitmq.pp

- Greatly simplified by allowing the rabbitmq module to do all of the heavy lifting for us

### manifests/profile/repos.pp

- Fixed a bug where we were pointing to an all lowercase URL which caused st2 package installs to fail

### manifests/profile/selinux.pp

- Added a class that configures SELinux on RHEL hosts

### manifests/profile/server.pp

- Small changes here related to adding database auth capability
- Added stanley user creation
- Added datastore crypto creation
- Added additional dependency management

### manifests/profile/web.pp

- Completely re-written
- I don't believe that st2web was complete when this module was last touched, so this class got a complete overhaul

### manifests/rbac.pp

- Fixed a few puppet lint errors
- Fixed an error where the RBAC rules were executed every puppet run

### manifests/server/datastore_keys.pp

- New manifest that manages the datastore crypto keys

### manifests/stanley.pp

- Removed unnecessary warning about ssh keys

### manifests/user.pp

- Fixed a couple small bugs related to a legacy "robots" group.
- This got a pretty big overhaul with regards to SSH key creation. Now, if SSH keys are not present new ones will be created (just like the shell scripts and  ansible-st2)

### metadata.json

- Reformatted the whole file to standard JSON formatting scheme 
- Updated module dependencies (some were missing)
- Added supported OS block
- Added supported puppet versions block

### spec/*

- Lots of small fixes here related to running the tests on various versions of ruby.
- Finally found a happy medium where all tests now pass

### CODE REVIEW

**Comment by nmaludy** [comment]
This fixes #159

**Comment by armab** [comment]
https://github.com/StackStorm/puppet-st2#installing-and-configuring-packs is outdated too.

**Comment by armab** [comment]
For some reason it changing the file permissions to `755` for all files in `/opt/stackstorm/packs`
```
Notice: /Stage[main]/St2::Profile::Fullinstall/St2::Pack[st2]/File[/opt/stackstorm/packs/core/actions/announcement.yaml]/mode: mode changed '0664' to '0775'
Notice: /Stage[main]/St2::Profile::Fullinstall/St2::Pack[st2]/File[/opt/stackstorm/packs/core/actions/remote.yaml]/mode: mode changed '0664' to '0775'
Notice: /Stage[main]/St2::Profile::Fullinstall/St2::Pack[st2]/File[/opt/stackstorm/packs/packs/actions/pack_mgmt/search.py]/mode: mode changed '0664' to '0775'
Notice: /Stage[main]/St2::Profile::Fullinstall/St2::Pack[st2]/File[/opt/stackstorm/packs/packs/actions/pack_mgmt/download.py]/mode: mode changed '0664' to '0775'
```

File: manifests/user.pp [hunk 1]
```diff
@@ -32,28 +29,25 @@
...
   })
 
   if $create_sudo_entry {
-    ensure_resource('sudo::conf', $_robots_group_name, {
+    ensure_resource('sudo::conf', $name, {
       'priority' => '10',
-      'content'  => "%${_robots_group_name} ALL=(ALL) NOPASSWD: SETENV: ALL",
+      # note: passes in $name variable into template
+      'content'  => template('st2/etc/sudoers.d/user.erb'),
     })
```
Comment (armab) [inline manifests/user.pp]: Under ubuntu xenial Vagrant, found that something breaks paswordless sudo for the default user after running the provisioner.
Will check later what's going on there.

File: manifests/pack.pp [hunk 2]
```diff
@@ -59,19 +79,26 @@
...
-  Package<| tag == 'st2::package::install' |> -> File['/opt/stackstorm/packs']
-  Service<| tag == 'st2::profile::service' |> -> Exec["install-st2-pack-${name}"]
-
   if $config {
     validate_hash($config)
-    file { "/opt/stackstorm/packs/${pack}/config.yaml":
+    file { "/opt/stackstorm/configs/${pack}.yaml":
       ensure  => file,
-      mode    => '0440',
+      mode    => '0755',
```
Comment (nmaludy) [inline manifests/pack.pp]: I was just going by what's present in my "one liner" install:

```
# ls -al /opt/stackstorm/configs/
total 16
drwxr-xr-x.  2 st2  root 109 Jul 27 11:27 .
drwxr-xr-x. 11 root root 204 Aug  2 21:46 ..
-rwxr-xr-x.  1 st2  root 148 Jun 30 10:42 activedirectory.yaml
-rwxr-xr-x.  1 st2  root  72 Jun 29 23:16 digitalocean.yaml
-rwxr-xr-x.  1 st2  root  17 Jun 13 09:40 testconfigdefault.yaml
-rwxr-xr-x.  1 st2  root 162 Jun  7 13:01 vsphere.yaml
```

File: manifests/params.pp [hunk 3]
```diff
@@ -190,10 +197,7 @@
         }
       } elsif $::operatingsystem == 'Ubuntu' {
         $init_type = $::operatingsystemmajrelease ? {
```
Comment (nmaludy) [inline manifests/params.pp]: No we don't, i'll removed this. 

There are a few other variables in this params class that are unused. I'll try to clean them up the best i can.

File: manifests/pack.pp (hunk 2 above)
Comment (armab) [inline manifests/pack.pp]: `+x` flag is not needed on yaml files, so `0640` should be fine.

**Comment by nmaludy** [comment]
@armab i still agree with your permissions decision, however i found the code in the base repo that sets these permissions to 775 https://github.com/StackStorm/st2/blob/master/contrib/packs/actions/pack_mgmt/download.py#L219-L238

For now i'll just ensure that the packs/ directory and all files under it are owned by group `st2packs`

**Comment by armab** [comment]
@nmaludy Thanks for the pointer! It turned into internal discussion with StackStorm team and https://github.com/StackStorm/st2/issues/3660 we think we should change that in st2 core.

File: manifests/user.pp (hunk 1 above)
Comment (armab) [inline manifests/user.pp]: So before running the `Puppet-st2`, I have the following:
```
ubuntu@ubuntu16:~$ sudo ls -la /etc/sudoers.d/
total 16
drwxr-x---  2 root root 4096 Aug  8 15:35 .
drwxr-xr-x 91 root root 4096 Aug  8 15:35 ..
-r--r-----  1 root root  123 Aug  8 15:35 90-cloud-init-users
-r--r-----  1 root root  958 Mar 30  2016 README
```

Running the Puppet provisioner removes that file, - that's why it breaks exsiting sudoers.

Comment (armab) [inline manifests/user.pp]: Guessing if Puppet's sudo module has some flag to leave previous sudo records in place.

**Comment by nmaludy** [comment]
@armab i fixed the problem with extra "unmanaged" `/etc/sudoers.d/*` files being purged during runtime.

This "purge prevention" will be the default behavior in this st2 module.

My change also tries to be smart and externally compatible for users that wish to configure the sudo package however they like (maybe they want the purge feature). In this case the code detects if the `::sudo` class has been instantiated previously and allows for the previous instance to dictate the main behavior.

**Pull request merged**